src/tsdb.c

```diff
@@ -214,23 +229,118 @@ size_t SeriesGetNumSamples(const Series *series) {
     return numSamples;
 }
 
+static void upsertRules(Series *series, AddCtx *aCtx) {
+    CompactionRule *rule = series->rules;
+    RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
+    while (rule != NULL) {
+        // upsert in latest timebucket
+        rule->backfilled = true;
+        if (aCtx->sample.timestamp >= CalcWindowStart(series->lastTimestamp, rule->timeBucket)) {
```

Comment:
For simplicity we probably want to recalculate the unflushed aggregation here and keep in the in `rule` context instead of pushing this logic to `handleCompaction`, it will also clean `handleCompaction` from any handling of OOO and will move all handling of the OO in 1 place and reduce the chances of bugs.

Reply:
Good idea